### ISSUE TITLE
Is it possible to build a custom PRIME_APP.DAT with your tools?

### ISSUE THREAD

**Issue by parisseb** [Project-Muteki/besta-tools#6]
Your project looks like very promising, I'd be very interested to know if it is possible to build KhiCAS for the Prime G1 using your tools, like ExistOS does for the hp39gii.
This would bring up some welcome enhancements on the G1 wrt the builtin CAS including:
* up to date Giac/Xcas version with all commands available
* inclusion of LGPL-ed libraries: GMP, MPFR, MPFI: faster bigint support, bigfloat library with arbitrary number of digits computation, interval arithmetic
* programming editor and shell with syntax coloration, parenthese match
* 3d geometry app
* javascript support
That way the Prime G1 would match other ports of KhiCAS on the Casio fxcg50, Numworks N0110/N0115/N0120 and TI Nspire.

**Comment by dogtopus** [comment]
First of all, thanks for your interest in this project. It's pretty rare to see others here.

Also, I'm not very familiar with Prime G1 since I don't own one and reversing the G1 firmware isn't really my priority. The priority was always running code on Besta dictionaries, since they have a less restrictive OS with official ways to load external binaries without reflashing boot (ROM) and system (DAT) partitions, unlike G1.

AFAIK almost all of the Prime G1 firmware logic resides in `armfir.elf`. It should be possible to write code that replaces it. Maybe you could use the GCC provided here to compile the code, but use a custom ldscript to place the ELF in the correct address. A second possible option could to be changing the startup INI file to use the relocatable PE files that `elf2bestape` would generate.

That being said, I'm not sure how beneficial this would really be comparing to something similar to ripem, but with a less janky port of an off-the-shelf OS instead (like FreeRTOS or a heavily trimmed Linux kernel), since AFAIK S3C2416 is pretty well documented.

**Comment by dogtopus** [comment]
Also to answer your question on the title:

> Is it possible to build a custom PRIME_APP.DAT with your tools?

No, at least not at the moment. You'd need to modify that image yourself. AFAIK it only has 1 section (also being a FAT16 partition that you can mount read/write and write to it) but there is some checksum involved, that I still didn't bother to take a look at how it really works.

**Comment by parisseb** [comment]
Switching to another OS is far out of reach for me. In addition, there is a risk to brick the device. I believe that modifying PRIME_APP.DAT is much more secure, if the device fails to boot, it's certainly possible to do a reset while keeping the Symb key pressed and reflash it with HP connectivity kit.
How did you mount the PRIME_APP.DAT file? I tried mount -o loop with a few -t options without success.

**Comment by parisseb** [comment]
After doing some hexedit, it seems there is a header of size 2K that does not change between different firmware versions, then the remaining part of the file can be mounted with -t msdos.

**Comment by dogtopus** [comment]
Yes. It's (mostly) in standard Besta data partition image format and there's a header that you need to either skip or cut out before you mount and edit the filesystem itself. The fact that the partition image has only 1 section made it a lot easier to edit as well.

**Comment by parisseb** [comment]
Once the fat16 partition is mounted, ls returns
ls /a
appslist.inf  appslist.map  appslist.md5  bestabfs.ind  firstrun.ini  programs  window

bestabfs.ind is empty. appslist.md5 is a list of MD5 signature.
more /a/firstrun.ini displays:
[RUNING SETTING]
FirstRunApp = \\.\ARMFIR.ROM

hexedit /a/appslist.inf has references to BESTAFIR.ROM/bestafir.exe, ARMFIR.ROM/armfir.elf, ARMFIRDB/armfir.dat, HELLO.ROM/hello.exe, ARMHELLO.ROM/armhello.elf, ARMHELLODB.ROM/armhello.dat, DIAGNOSE.ROM/diagnose.exe, 

appslist.map is a binary file, it begins by 7, which is the number of references above. I don't know what follows, but I'm afraid it should be modified if the partition is modified.

ls /a/programs/
misc  tools
ls /a/programs/tools
bestafir.dat  bestafir.exe
ls /a/programs/misc
armfir.dat  armfir.elf  diagnose.dat  diagnose.exe

It is perhaps possible to add an elf file in /programs/misc that is already referenced, e.g. armhello.elf, and replace armfir.rom with armhello.rom in firstrun.ini. This might be better than replacing armfir.elf, because maybe one can load armfir.elf later from hello, if the user want to run the calculator firmware.

**Comment by parisseb** [comment]
Commandline to get the partition from PRIME_APP.DAT
dd if=PRIME_APP.DAT of=fat16 skip=1 bs=8192
Mounting to an existing /path
sudo mount -t msdos -o loop fat16 /path

**Comment by parisseb** [comment]
Getting the header
dd if=PRIME_APP.DAT of=header count=1 bs=8192
Merge header with partition
/bin/cp header FILE.DAT && dd if=fat16 of=FILE.DAT oflag=append conv=notrunc  bs=8192

**Comment by dogtopus** [comment]
I'll move this to discussions since it's not directly related to the tools themselves.
